### PR TITLE
Improve UX of send and status commands

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Improve shared context utility with conditional typing and core artifact handling
- Improve UX of send and status commands with prompts
- Prompt for key if required and not provided

### Drive-by Changes

Update the token readme for https://github.com/hyperlane-xyz/issues/issues/715

### Backward compatibility

Yes

### Testing

Manual

### CODE REVIEW

**Comment by jmrossy** [comment]
More updates for https://github.com/hyperlane-xyz/issues/issues/713

**Comment by avious00** [comment]
thanks for adding in the context / coreartifact section, LGTM

File: typescript/cli/src/commands/send.ts [hunk 1]
```diff
@@ -63,9 +61,9 @@ const messageCommand: CommandModule = {
   handler: async (argv: any) => {
     const key: string = argv.key || process.env.HYP_KEY;
     const chainConfigPath: string = argv.chains;
-    const coreArtifactsPath: string = argv.core;
-    const origin: string = argv.origin;
-    const destination: string = argv.destination;
+    const coreArtifactsPath: string | undefined = argv.core;
+    const origin: string | undefined = argv.origin;
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/cli/src/commands/send.ts:65]: No, as part of these changes, they are now optional and the command prompt for their section from the chain list as needed

File: typescript/cli/src/commands/send.ts [hunk 2]
```diff
@@ -114,11 +112,11 @@ const transferCommand: CommandModule = {
...
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/cli/src/commands/send.ts:116]: same as above :)

File: typescript/cli/src/context.ts [hunk 3]
```diff
@@ -43,17 +45,69 @@ export function getMergedContractAddresses(
...
+  multiProvider: MultiProvider;
 }
 
-export function getContextWithSigner(key: string, chainConfigPath: string) {
-  const signer = keyToSigner(key);
+// This makes return type dynamic based on the input settings
+type CommandContext<P extends ContextSettings> = CommandContextBase &
+  (P extends { keyConfig: object }
+    ? { signer: ethers.Signer }
+    : { signer: undefined }) &
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/cli/src/context.ts:69]: Some commands, like the config and status commands, don't require a signer.
This is a fancy type that returns defined/undefined signer and/or core depending on the params given

File: typescript/cli/src/context.ts [hunk 4]
```diff
@@ -43,17 +45,69 @@ export function getMergedContractAddresses(
...
+  let signer = undefined;
+  if (keyConfig) {
+    const key =
+      keyConfig.key ||
+      (await input({
+        message:
+          keyConfig.promptMessage ||
+          'Please enter a private key or use the HYP_KEY environment variable',
+      }));
+    signer = keyToSigner(key);
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/cli/src/context.ts:90]: If is part of `new ethers.Wallet`

File: typescript/cli/src/send/message.ts [hunk 5]
```diff
@@ -32,16 +27,32 @@ export async function sendTestMessage({
...
-    ? readDeploymentArtifacts(coreArtifactsPath)
-    : undefined;
+  const { signer, multiProvider, customChains, coreArtifacts } =
+    await getContext({
+      chainConfigPath,
+      coreConfig: { coreArtifactsPath },
+      keyConfig: { key },
+    });
+
+  if (!origin) {
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/cli/src/send/message.ts:43]: See above

**Review by aroralanuk — APPROVED** [review]

**Pull request merged**